### ISSUE TITLE
Issues with function_ref

### ISSUE THREAD

**Issue by ymarcov** [eclipse-iceoryx/iceoryx#1897]
## Required information

**Operating system:**
Any

**Compiler version:**
Any

**Eclipse iceoryx version:**
master

**Observed result or behaviour:**
In `cxx::function_ref`:
1. `operator()` is marked `noexcept`
2. Construction erroneously accepts rvalue capturing lambdas (i.e. `[&] { ... }`)
3. Construction invokes UB when accepting rvalue non-capturing lambdas (i.e. `[] { ... }`) 

**Expected result or behaviour:**
1. `operator()` should not be marked `noexcept`, since it is unreasonable for the program to terminate if the user-code invoked throws an exception.
2. Capturing lambdas necessarily imply state, and it is a mistake that `function_ref` can be built with stateful rvalues. Attempting this construction should result in a compilation error.
3. Non-capturing lambdas should be okay for construction, since they are convertible to a function pointer. However, the current implementation still saves them as objects, and applies `callable_object(args...)` on them where `callable_object` is now necessarily a dangling reference.

**Comment by elfenpiff** [comment]
@ymarcov Thanks for the finding.

1. I think we can easily solve this by removing `noexcept`. Introducing `noexcept` was one of our early mistakes which we really have to face this year.

Could you elaborate 2 a bit more in detail. The contract is that a `function_ref` does not own the callable and the user has to ensure the lifetime of it by other means. If you require an owning function you have to use `function`.
So a capturing lambda, or lambda with with state should be alright as long as it lifes at least as long as the `function_ref`. But I assume that you meaning constructs like (A):
```cxx
function_ref f = [&]() { doStuff(capturedVariable); };
// the lambda is a temporary rvalue and the lifetime is shorter than f and we have a problem 
```

But in a construct like this it would be alright (B):
```cxx
void callMe(const function_ref<void()> &f) {
  f();
} 

callMe([&]() { doStuff(capturedVariable);});
```
Here the C++ standard guarantees that the lambda is a temporary rvalue variable which lifes until the end of the statement and the call to the function is finished.

This would in summary mean that when we fix 2. we wouldn't be able to create a `function_ref` like in (A) which would prevent bugs but also our major use case (B) wouldn't work anymore and for iceoryx `function_ref` would maybe become useless.

If this is correct I would tend to remove `function_ref` completely since it is too easy to shoot yourself in the face and capturing lambdas is for me a must have otherwise I do not see any advantage over a simple and direct c function ref as argument.

**Comment by ymarcov** [comment]
@elfenpiff I think that it is indeed risky to be able to construct an invalid `function_ref`, especially in so easy a manner.

For your `callMe()` example, I'd probably replace that with 

```c++
template <typename Func>
void callMe(Func&& f);
```

If the user didn't want to store this function-object beyond the call, it would still be legit to use like that. And if he did, then he'd have to store it in a normal `function`, which would be valid.

Re the removal of `function_ref`, hard for me to say since this is a useful construct, but if other than the use case you mentioned there is no real use for it in this codebase, then perhaps it could be removed.

**Comment by mossmaurice** [comment]
@ymarcov Thanks for creating this issue!

1. I agree with what was said above.
2. I understand the issue and that it can . However, my point is that with similar non-owning construct the same issue exists (e.g. `std::string_view`, `std::span`, etc.). It was even considered in the [proposal from 2019 in section 17.2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0792r5.html#possible-issues). The latest paper [presents a survey](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p0792r10.html#Survey) on implementation in different base libraries. E.g. in Abseil a `function_ref` is not assignable to [prevent subtle lifetime bugs](https://github.com/abseil/abseil-cpp/blob/master/absl/functional/function_ref.h#L125-L128). So it might be worth looking at the other implementations mentioned there. I would refrain from removing the `function_ref` completely as non-owning wrapper like `std::span` and friends are very useful in embedded, real-time development. What we could discuss is to remove `function_ref` from the public API.
3. The solution could be to always use [this c'tor](https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp#L94) for everything which is convertible to a function pointer. Need to think more about it, though.

cc @elfenpiff

**Comment by ymarcov** [comment]
@mossmaurice 

1. Check.
2. Good argument IMO. One could accept a `function_ref` only when one is sure this won't be stored, otherwise (by default, maybe) accept a `function`.
3. Good solution IMO.

**Comment by elBoberido** [comment]
@elfenpiff well, I was aware* that storing a `function_ref` has issues and I always advocated to only use is as parameter in functions but never store it. We also mused a few times about the `*_view` classes of the STL, why I think they are a source for bugs and that they should only be used as "transport" type by not as "storage" type. Saying that, `function_ref` has it's use cases and as long as we can prevent a bug-prone usage like storing a `function_ref` I would keep it. Stuff like
```cpp
function_ref f{[&]() { doStuff(capturedVariable); }};
std::move(f)();
```
I do not consider bug prone since it a deliberate decision of the user but allowing only rvalues would probably solve our issues.

Btw, it looks like we pass a reference of a `function_ref` as parameter at some places.

cc @ymarcov @mossmaurice 

*) this does not mean that I do not make the mistake myself to accidentally recommend the use of a `function_ref` as template parameter which accepts any callable and stores that parameter like in the `PeriodicTask` 🙈

**Comment by ymarcov** [comment]
@elBoberido just to make sure we're all on the same page here, the following is necessarily a bug given the concept of function ref, do we all agree on this?

```c++
function_ref<void()> f{[&] { }};
// Invoking f() is always a bug now
```

**Comment by ymarcov** [comment]
My feeling here is that it isn't possible to provide function_ref in a way that is both elegantly usable, as well as foolproof. Same as _view classes.

I would add here, though, that unlike string_view and friends which provide a non trivial and uniform API, I'm not sure that if ou use function ref only for "transition" objects, it is better than a templated Callable. Except, it could be useful for virtual functions which cannot take a template, but from what I understand this is not a viable use case for you.

**Comment by ymarcov** [comment]
I liked the idea of @mossmaurice to allow this constructor, but provide a lint rule to check it is used correctly.

**Comment by mossmaurice** [comment]
> @elfenpiff We weren't aware of the issue until @ymarcov discovered it and I would bet it would also slip future PR.
> 
> @elBoberido well, I was aware* that storing a function_ref has issues and I always advocated to only use is as parameter in functions but never store it.

That's not correct, we were aware see [the doxygen](https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp#L60-L64).

Back during the first implementation, I agreed and still agree with what is written in the [proposal](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0792r5.html):

> I strongly believe that accepting temporaries is a “necessary evil” for both function_ref and std::string_view, as it enables countless valid use cases. The problem of dangling references has been always present in the language - a more general solution like Herb Sutter and Neil Macintosh’s lifetime tracking18 would prevent mistakes without limiting the usefulness of view/reference classes.

IMHO we can't change the pitfalls of the core C++ language without switching to other more modern languages like Rust, which offer concepts like borrowing.

> With these arguments in mind I would remove function_ref.

If we apply your point, this would also mean `std::span` and `std::string_view` and friends should be avoided. However, those actually lead to safer code to prevent API calls like

`copyBuffer(char* buffer, size_t size)`

with

`copyBuffer(std::span buffer)`

but of course they can also be misused when using them not as function arguments, but to stored something.

> My feeling here is that it isn't possible to provide `function_ref` in a way that is both elegantly usable, as well as foolproof.

Full ack :100: 

> provide a lint rule to check it is used correctly.

I like the idea! What I would propose is to write a custom-clang tidy rule which checks that those constructs are only used as function parameter. This could be even accepted upstream for `string_view` and friends. Maybe it already exists. Such a rule and an address sanitizer to detect invalid memory access should provide enough safety to use them. Does that make sense?

cc @ymarcov
